File: SortingAlgorithms/Merge Sort/MergeSort.py
"""
Author: Ao Wang
Date: 08/20/19
Description: Merge Sort Algorithm
"""

def merge(left, right):
    result = []
    leftInd = 0
    rightInd = 0
    resultInd = 0
    
    while leftInd < len(left) or rightInd < len(right):
        if leftInd < len(left) and rightInd < len(right):
            if left[leftInd] < right[rightInd]:
                result.append(left[leftInd])
                leftInd += 1
            else:
                result.append(right[rightInd])
                rightInd += 1
        elif leftInd < len(left):
            result.append(left[leftInd])
            leftInd += 1
        elif rightInd < len(right):
            result.append(right[rightInd])
            rightInd += 1
    
    return result

def mergeSort(nums):
    if len(nums) == 1:
        return nums
    
    mid = len(nums) // 2
    left = []
    right = []
    
    for i in range(mid):
        left.append(nums[i])
    for i in range(mid, len(nums)):
        right.append(nums[i])
    
    left = mergeSort(left)
    right = mergeSort(right)
    
    return merge(left, right)

nums = [0, -1, 100, 5, -99, 85, 90, 3]
print("Unsorted: " + str(nums))
print("Sorted: " + str(mergeSort(nums)))

File: SortingAlgorithms/Merge Sort/MergeSort.java
import java.util.Arrays;

public class MergeSort {

	public static void main(String[] args) {

		int[] nums = { 0, -1, 100, 5, -99, 85, 90, 3 };
		System.out.println("Unsorted: " + Arrays.toString(nums));
		System.out.println("Sorted: " + Arrays.toString(mergeSort(nums)));
	}

	public static int[] mergeSort(int[] nums) {

		if (nums.length == 1) {
			return nums;
		}

		int mid = nums.length / 2;
		int[] left = new int[mid];
		int[] right = new int[nums.length - mid];

		for (int i = 0; i < left.length; i++) {
			left[i] = nums[i];
		}
		for (int i = 0; i < right.length; i++) {
			right[i] = nums[mid + i];
		}

		left = mergeSort(left);
		right = mergeSort(right);

		return merge(left, right);

	}

	public static int[] merge(int[] left, int[] right) {

		int[] result = new int[left.length + right.length];

		int leftInd = 0;
		int rightInd = 0;
		int resultInd = 0;

		while (leftInd < left.length || rightInd < right.length) {
			if (leftInd < left.length && rightInd < right.length) {
				if (left[leftInd] < right[rightInd]) {
					result[resultInd++] = left[leftInd++];
				} else {
					result[resultInd++] = right[rightInd++];
				}
			} else if (leftInd < left.length) {
				result[resultInd++] = left[leftInd++];
			} else if (rightInd < right.length) {
				result[resultInd++] = right[rightInd++];
			}
		}

		return result;
	}
}

